### PR TITLE
Refactor to provide infrastructure for Multi-agent environments

### PR DESCRIPTION
**Pull request by lqy0057**
## What changes do you make in this PR?

* Please describe why you create this PR

## Checklist

* [ ] I have merged the latest main branch into current branch.
* [ ] I have run `bash scripts/format.sh` before merging.
* Please use "squash and merge" mode.

### CODE REVIEW

**Comment by lqy0057** [comment]
close #273

**Comment by lqy0057** [comment]
Close #273

File: pgdrive/envs/pgdrive_env.py [hunk 1]
```diff
@@ -287,33 +227,23 @@ def step(self, actions: Union[np.ndarray, Dict[AnyStr, np.ndarray]]):
...
+            vehicle.step_info["step_reward"] = float(reward)
             rewards[key] = reward + done_reward
 
         if self.num_agents == 1:
             return obses[DEFAULT_AGENT], rewards[DEFAULT_AGENT], \
-                   copy.deepcopy(self.dones[DEFAULT_AGENT]), copy.deepcopy(step_infos[DEFAULT_AGENT])
+                   copy.deepcopy(self.dones[DEFAULT_AGENT]), copy.deepcopy(self.vehicle.step_info)
         else:
-            return obses, rewards, copy.deepcopy(self.dones), copy.deepcopy(step_infos)
+            return obses, rewards, copy.deepcopy(self.dones), self.for_each_vehicle(lambda v:v.step_info)
```
Comment (pengzhenghao) [inline pgdrive/envs/pgdrive_env.py:246]: add copy.deepcopy

File: pgdrive/envs/pgdrive_env.py [hunk 2]
```diff
@@ -812,11 +597,19 @@ def for_each_vehicle(self, func, *args, **kwargs):
...
         assert len(self.vehicles) == 1, "env.vehicle is only supported in single-agent environment!"
-        return self.vehicles[DEFAULT_AGENT]
+        ego_v = self.vehicles[DEFAULT_AGENT]
+        return ego_v
 
     def reward(self, *args, **kwargs):
         raise ValueError("reward function is deprecated!")
 
+    @property
+    def use_image(self):
```
Comment (pengzhenghao) [inline pgdrive/envs/pgdrive_env.py:607]: The name is confusing, change to use_render or use_camera or use_first_view or so.

File: pgdrive/tests/test_pgdrive_env.py [hunk 3]
```diff
@@ -86,8 +86,8 @@ def test_zombie():
             acc = acc_controller.get_result(acc_error)
             if d:
                 assert info["arrive_dest"]
-                assert abs(env.vehicles[env.DEFAULT_AGENT].position[0] - dest[0]) < 0.1 and \
-                       abs(env.vehicles[env.DEFAULT_AGENT].position[1] - dest[1]) < 0.1
+                assert abs(env.vehicles[env.DEFAULT_AGENT].position[0] - dest[0]) < 0.15 and \
+                       abs(env.vehicles[env.DEFAULT_AGENT].position[1] - dest[1]) < 0.15
```
Comment (pengzhenghao) [inline pgdrive/tests/test_pgdrive_env.py:90]: Are these changes under expectation?

**Comment by lqy0057** [comment]
solve #263

File: pgdrive/tests/test_pgdrive_env.py (hunk 3 above)
Comment (lqy0057) [reply pgdrive/tests/test_pgdrive_env.py:90]: yes, only 1e-2 m difference comparing to the previous edition

**Pull request merged**